### PR TITLE
Fix: Fix Stats and Utxo cache

### LINKED ISSUE #11: Stale block address indexing bug
Electrs appears to double-count certain transactions during address indexing, which produces incorrect chain stats (`funded_txo_sum` etc) and therefore incorrect balances displayed on mempool.space.

Screenshot 2023-06-05 at 11 39 08 AM

---

Specifically, the error seems to occur when a transaction is included in a stale block, but *not* the "winning" block at the same height.

For example, transaction [3738...e88e](https://mempool.space/tx/3738b0795522795d50fe9c385154abbb41585d425a402c9979db20758ba9e88e
)
was included in a recent stale block 000000000000000000032956cbfd8721abe0572d81542e75e38e6185312bee09 at height 792379, but eventually entered the main chain in block [792380](https://mempool.space/block/792380).

The transaction is then double-counted during address indexing, 
causing address [132kFhZzkFytWAefWy7nCYgJHcU84RjJNa
](https://mempool.space/address/132kFhZzkFytWAefWy7nCYgJHcU84RjJNa
) to show a negative balance.

Similarly, transaction [3264...d9cb](https://mempool.space/tx/326458888709874443396dc5f66eb359cd59e7984dcc0778b64d126cf0b2d9cb) appears in the same stale block, enters the main chain in block [792382](https://mempool.space/block/792382), and is counted twice towards address [bc1qzuupxjkwv4dj69m3xzack5sdqtacn6l8czndlf](https://mempool.space/address/bc1qzuupxjkwv4dj69m3xzack5sdqtacn6l8czndlf), causing a positive balance to display when in fact the address should be empty.

Screenshot 2023-06-05 at 12 05 33 PM

---

Transactions which appear at the *same* height on both sides of a chain split appear to be handled correctly.

For example, address [bc1pfajpg0fp0a62wjrax682hkzvrczxx7j3htrtm0ae6cnx72ye2rnscarxrp](https://mempool.space/address/bc1pfajpg0fp0a62wjrax682hkzvrczxx7j3htrtm0ae6cnx72ye2rnscarxrp) has a transaction which is included in both versions of block 792379, but reports the correct balance.

### PR DESCRIPTION
**Pull request by junderw**
Fixes #11 

At the very least it fixes the more embarrassing ones.

1. Negative balance
2. Non-zero balance even though number of funds and spends txos are equal etc.

Also, comparing the confirmed_height in the history key to the blockid height (from the local best header chain) is a more efficient way to weed out the re-confirmed-in-separate-block issue.
If a re-org happens and the same TX is in the same height block, then it will perfectly match the existing key (and won't duplicate)

### CODE REVIEW

File: src/new_index/schema.rs [hunk 1]
```diff
@@ -145,6 +145,29 @@ pub struct ScriptStats {
...
+    }
+    #[cfg(not(feature = "liquid"))]
+    fn is_sane(&self) -> bool {
+        // There are less or equal spends to funds
+        self.spent_txo_count <= self.funded_txo_count
+        // There are less or equal transactions to total spent+funded txo counts
+        // (Most spread out txo case = N funds in 1 tx each + M spends in 1 tx each = N + M txes)
+        && self.tx_count <= self.spent_txo_count + self.funded_txo_count
+        // There are less or equal spent coins to funded coins
+        && self.spent_txo_sum <= self.funded_txo_sum
```
Comment (junderw) [inline src/new_index/schema.rs:163]: This prevents negative balance.

File: src/new_index/schema.rs [hunk 2]
```diff
@@ -145,6 +145,29 @@ pub struct ScriptStats {
...
+        // If funded and spent txos are equal (0 balance)
+        // Then funded and spent coins must be equal (0 balance)
+        && (self.funded_txo_count == self.spent_txo_count)
+            == (self.funded_txo_sum == self.spent_txo_sum)
```
Comment (junderw) [inline src/new_index/schema.rs:167]: This prevents positive balance when there is an equal number of fund and spend txos.

ie. The balance should be 0 but some the balance value is some small amount like a few satoshis.

**Review by mononaut — APPROVED** [review]
utACK @ a824fd8e

I haven't been able to find any bad stats locally, or reproduce the issue in regtest, so I couldn't test the actual fix.

but the code looks good, and it doesn't seem to break anything.

**Review by softsimon — APPROVED** [review]
utACK

**Pull request merged**